authentication.go

```diff
@@ -0,0 +1,83 @@
+package resource
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+type InstallationResponse struct {
+	Id int
+}
+
+type TokenResponse struct {
+	Token string
+}
+
+func GenerateAccessToken(s *Source, now time.Time) (string, error) {
+	if s.AccessToken != "" {
+		return s.AccessToken, nil
+	}
+
+	decode, _ := pem.Decode([]byte(s.PrivateKey))
+	key, _ := x509.ParsePKCS1PrivateKey(decode.Bytes)
+	sig, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.RS256, Key: key}, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		panic(err)
+	}
+
+	cl := jwt.Claims{
+		Issuer:   s.AppId,
+		IssuedAt: jwt.NewNumericDate(now),
+		Expiry:   jwt.NewNumericDate(now.Add(9 * time.Minute)),
+	}
+	signedJwt, err := jwt.Signed(sig).Claims(cl).CompactSerialize()
+	if err != nil {
+		panic(err)
+	}
+
+	var endpoint string
+	if s.V3Endpoint != "" {
+		endpoint = s.V3Endpoint
+	} else {
+		endpoint = "https://api.github.com"
+	}
+
+	installationResponse := callApi("GET", endpoint+"/repos/"+s.Repository+"/installation", signedJwt)
+	var ir InstallationResponse
+	err = json.NewDecoder(installationResponse.Body).Decode(&ir)
+	if err != nil {
+		panic(err)
+	}
+
+	tokenResponse := callApi("POST", endpoint+"/app/installations/"+strconv.Itoa(ir.Id)+"/access_tokens", signedJwt)
```

Comment:
No not really, it's just best practice in Go to close an HTTP response body once you're done with it so-as to not leak the connection. In the context of Concourse, it probably does not matter as much as it would in a long running HTTP server.

Reply:
No not really, it's just best practice in Go to close an HTTP response body once you're done with it so-as to not leak the connection. In the context of Concourse, it probably does not matter as much as it would in a long running HTTP server.